File: flutter-wrapper/lib/proxy/domain_resolver.dart
import 'dart:async';

import 'package:flutter/foundation.dart';
import 'package:http/http.dart' as http;

class DomainResolver {
  static const Duration _fetchTimeout = Duration(seconds: 5);
  static const Duration _probeTimeout = Duration(seconds: 3);

  /// Try to swap the host of [originalUrl] for the first reachable domain
  /// listed in any of [domainConfigUrls]. Returns [originalUrl] if no override
  /// can be resolved or none of the candidates respond.
  ///
  /// Each entry in [domainConfigUrls] points at a plain-text file that lists
  /// one candidate domain per line. We fetch the first config URL that
  /// succeeds, then probe each domain in order with HEAD (falling back to GET)
  /// and pick the first 2xx/3xx response.
  static Future<Uri> resolve({
    required Uri originalUrl,
    required List<String> domainConfigUrls,
  }) async {
    if (domainConfigUrls.isEmpty) return originalUrl;

    final candidates = await _fetchCandidates(domainConfigUrls);
    if (candidates.isEmpty) return originalUrl;

    for (final domain in candidates) {
      final candidate = originalUrl.replace(host: domain, port: 0);
      if (await _isReachable(candidate)) {
        debugPrint('DomainResolver: selected $candidate');
        return candidate;
      }
    }

    debugPrint('DomainResolver: no candidate reachable, falling back to $originalUrl');
    return originalUrl;
  }

  static Future<List<String>> _fetchCandidates(List<String> configUrls) async {
    for (final raw in configUrls) {
      final uri = Uri.tryParse(raw);
      if (uri == null) continue;
      try {
        final response = await http.get(uri).timeout(_fetchTimeout);
        if (response.statusCode >= 200 && response.statusCode < 300) {
          final lines = response.body
              .split(RegExp(r'\r?\n'))
              .map((l) => l.trim())
              .where((l) => l.isNotEmpty && !l.startsWith('#'))
              .toList();
          if (lines.isNotEmpty) return lines;
        }
      } catch (e) {
        debugPrint('DomainResolver: fetch $uri failed: $e');
      }
    }
    return const [];
  }

  static Future<bool> _isReachable(Uri url) async {
    try {
      final head = await http.head(url).timeout(_probeTimeout);
      if (_isOk(head.statusCode)) return true;
      // Some hosts disallow HEAD; fall back to GET.
      if (head.statusCode == 405 || head.statusCode == 501) {
        final get = await http.get(url).timeout(_probeTimeout);
        return _isOk(get.statusCode);
      }
      return false;
    } catch (e) {
      debugPrint('DomainResolver: probe $url failed: $e');
      return false;
    }
  }

  static bool _isOk(int status) => status >= 200 && status < 400;
}

File: electron-wrapper/proxy/domain-resolver.js
'use strict';

const FETCH_TIMEOUT_MS = 5000;
const PROBE_TIMEOUT_MS = 3000;

async function resolveReachableUrl({ originalUrl, domainConfigUrls }) {
  if (!Array.isArray(domainConfigUrls) || domainConfigUrls.length === 0) {
    return originalUrl;
  }

  const candidates = await fetchCandidates(domainConfigUrls);
  if (candidates.length === 0) return originalUrl;

  const original = new URL(originalUrl);
  for (const domain of candidates) {
    const candidate = new URL(originalUrl);
    candidate.host = domain;
    if (await isReachable(candidate.toString())) {
      console.log('domain-resolver: selected', candidate.toString());
      return candidate.toString();
    }
  }

  console.log('domain-resolver: no candidate reachable, falling back to', original.toString());
  return originalUrl;
}

async function fetchCandidates(configUrls) {
  for (const url of configUrls) {
    try {
      const response = await fetchWithTimeout(url, { method: 'GET' }, FETCH_TIMEOUT_MS);
      if (!response.ok) continue;
      const text = await response.text();
      const lines = text
        .split(/\r?\n/)
        .map((l) => l.trim())
        .filter((l) => l.length > 0 && !l.startsWith('#'));
      if (lines.length > 0) return lines;
    } catch (e) {
      console.warn('domain-resolver: fetch', url, 'failed:', e.message);
    }
  }
  return [];
}

async function isReachable(url) {
  try {
    const head = await fetchWithTimeout(url, { method: 'HEAD' }, PROBE_TIMEOUT_MS);
    if (head.status >= 200 && head.status < 400) return true;
    if (head.status === 405 || head.status === 501) {
      const get = await fetchWithTimeout(url, { method: 'GET' }, PROBE_TIMEOUT_MS);
      return get.status >= 200 && get.status < 400;
    }
    return false;
  } catch (e) {
    console.warn('domain-resolver: probe', url, 'failed:', e.message);
    return false;
  }
}

async function fetchWithTimeout(url, options, timeoutMs) {
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), timeoutMs);
  try {
    return await fetch(url, { ...options, signal: controller.signal });
  } finally {
    clearTimeout(timer);
  }
}

module.exports = { resolveReachableUrl };
